torch/csrc/api/include/torch/optim/serialize.h

```diff
@@ -40,6 +40,101 @@ void serialize(
     const std::string& key,
     std::vector<int64_t>& steps);
 
+template <typename DerivedOptimizerParamState>
+void serialize(
+    serialize::OutputArchive& archive,
+    const std::string& key,
+    const ska::flat_hash_map<std::string, std::unique_ptr<OptimizerParamState>>& state) {
+
+  serialize::OutputArchive state_archive;
+  for (const auto& item : state) {
+    serialize::OutputArchive param_state_archive; // For each OptimizerParamState
+    std::string tensorimpl_key = item.first;
+    const DerivedOptimizerParamState& curr_state = static_cast<const DerivedOptimizerParamState&>(*(item.second.get()));
+    curr_state.serialize(param_state_archive);
+    state_archive.write(tensorimpl_key, param_state_archive);
+  }
+  archive.write(key, state_archive);
+}
+
+template <typename DerivedOptimizerParamState>
+void serialize(
+    serialize::InputArchive& archive,
+    const std::string& key,
+    ska::flat_hash_map<std::string, std::unique_ptr<OptimizerParamState>>& state) {
+
+  serialize::InputArchive state_archive;
+  archive.read(key, state_archive);
+  std::vector<std::string> tensorimpl_keys = state_archive.keys();
+  for (const std::string& tensorimpl_key : tensorimpl_keys) {
+    serialize::InputArchive param_state_archive;
+    archive.read(tensorimpl_key, param_state_archive);
+    DerivedOptimizerParamState param_state;
+    param_state.serialize(param_state_archive);
+    state[tensorimpl_key] = c10::guts::make_unique<DerivedOptimizerParamState>(param_state);
```

Comment:
here, should we just have state as ska::flat_hash_map<std::string, DerivedOptimizerParamState>
and store its unique_ptr only when we are finally updating state in the load function because if this function is evaluated on a different machine then our stored ptr will be of no use ...

Reply:
ah I see, do you also suggest that we should change the signature of this function to take in `ska::flat_hash_map<std::string, DerivedOptimizerParamState>` instead? I am not entirely sure if it’s doable, because `state` is a non-const reference to the `state` field stored in OptimizerBase, which has type `ska::flat_hash_map<std::string, std::unique_ptr<OptimizerParamState>>`

Reply:
also this is the load function (which takes an InputArchive as parameter)

Reply:
yeah I was definitely suggesting that we should take in ska::flat_hash_map<std::string, DerivedOptimizerParamState> as an argument. if you look at the rough draft of the load function in Adagrad.cpp file, I have done something similar for param_groups and I think, similarly, we can create a local variable of the type ska::flat_hash_map<std::string, DerivedOptimizerParamState>, pass it to this function, and then update the state_ member variable in load